sycl/plugins/native_cpu/pi_native_cpu.cpp

```diff
@@ -0,0 +1,1308 @@
+    //               PI_ERROR_INVALID_ARG_VALUE);
+    return PI_ERROR_PLUGIN_SPECIFIC_ERROR;
+  }
+  default:
+    __SYCL_PI_HANDLE_UNKNOWN_PARAM_NAME(param_name);
+  }
+
+  return PI_ERROR_OUT_OF_RESOURCES;
+}
+
+pi_result piextContextSetExtendedDeleter(pi_context,
+                                         pi_context_extended_deleter, void *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piextContextGetNativeHandle(pi_context, pi_native_handle *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piextContextCreateWithNativeHandle(pi_native_handle, pi_uint32,
+                                             const pi_device *, bool,
+                                             pi_context *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piContextRetain(pi_context Context) { DIE_NO_IMPLEMENTATION; }
+
+pi_result piContextRelease(pi_context Context) {
+  // Todo: is it fine as a no-op?
+  return PI_SUCCESS;
+}
+
+pi_result piQueueCreate(pi_context Context, pi_device Device,
+                        pi_queue_properties Properties, pi_queue *Queue) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piQueueGetInfo(pi_queue, pi_queue_info, size_t, void *, size_t *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piQueueRetain(pi_queue Queue) { DIE_NO_IMPLEMENTATION; }
+
+pi_result piQueueRelease(pi_queue Queue) {
+  // Todo: is it fine as a no-op?
+  return PI_SUCCESS;
+}
+
+pi_result piQueueFinish(pi_queue) {
+  // Todo: is it fine as a no-op?
+  return PI_SUCCESS;
+}
+
+pi_result piQueueFlush(pi_queue) { DIE_NO_IMPLEMENTATION; }
+
+pi_result piextQueueGetNativeHandle(pi_queue, pi_native_handle *, int32_t *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piextQueueCreateWithNativeHandle(pi_native_handle, int32_t,
+                                           pi_context, pi_device, bool,
+                                           pi_queue_properties *, pi_queue *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
+                            void *HostPtr, pi_mem *RetMem,
+                            const pi_mem_properties *properties) {
+  // TODO: add proper error checking and double check flag semantics
+  // TODO: support PI_MEM_FLAGS_HOST_PTR_ALLOC flag
+  const bool UseHostPtr = Flags & PI_MEM_FLAGS_HOST_PTR_USE;
+  const bool CopyHostPtr = Flags & PI_MEM_FLAGS_HOST_PTR_COPY;
+  const bool HostPtrNotNull = HostPtr != nullptr;
+  if (UseHostPtr && HostPtrNotNull) {
+    *RetMem = new _pi_mem(HostPtr);
+  } else if (CopyHostPtr && HostPtrNotNull) {
+    *RetMem = new _pi_mem(HostPtr, Size);
+  } else {
+    *RetMem = new _pi_mem(Size);
+  }
+  return PI_SUCCESS;
+}
+
+pi_result piMemGetInfo(pi_mem, pi_mem_info, size_t, void *, size_t *) {
+  DIE_NO_IMPLEMENTATION;
+}
+
+pi_result piMemRetain(pi_mem Mem) { DIE_NO_IMPLEMENTATION; }
+
+pi_result piMemRelease(pi_mem Mem) {
+  Mem->RefCount.fetch_sub(1, std::memory_order_acq_rel);
+  if (Mem->RefCount == 0) {
```

Comment:
This is a race-condition. You could have two threads decrement the counter before either gets to the ref-count check below, which would result in a double free if the final ref-count is 0. Easy to fix though!

```suggestion
  uint32_t OldRefCount = Mem->RefCount.fetch_sub(1, std::memory_order_acq_rel);
  if (OldRefCount  == 1) {
```